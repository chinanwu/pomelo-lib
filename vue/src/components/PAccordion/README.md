# PAccordion

A base accordion component PAccordion.

## Example
```html
<PAccordion>
    <template #content>
        I am the accordion content!
    </template>
</PAccordion>
```
Alternatively, you can use the hyphenated component name.
```html
<p-accordion></p-accordion>
```

## Props

| Prop | Type |  Default | Description | Validation |
| ---- | ---- | -------- | ----------- | ---------- |
| label | String | Null | The string that will be displayed on the Accordion header | None |
| pClass | String | Null | Adds an additional class onto the accordion | None |
| expand | String | Null | Sets if the accordion is opened by default | None |

## Best practices

- TBA?

## Related components

- PButton: The accordion's header utilizes PButton
