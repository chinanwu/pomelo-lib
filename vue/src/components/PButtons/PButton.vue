<template>
  <button
    :id="id"
    class="PButton"
    :ariaLabel="ariaLabel"
    :ariaLabelledby="ariaLabelledBy"
    :role="role"
    :ariaHaspopup="ariaHasPopup"
    :ariaRoledescription="ariaRoledescription"
    :disabled="disabled"
    @click="handleClick"
    @keydown="handleKeyDown"
  >
    <slot name="content"></slot>
  </button>
</template>

<script>
// Basic Pomelo Button Component

export default {
  name: "PButton",
  props: {
    id: {
      type: String,
      default: null
    },
    ariaLabel: {
      type: String,
      default: null
    },
    ariaLabelledBy: {
      type: String,
      default: null
    },
    role: {
      type: String,
      default: null,
      validator: function(value) {
        return ["tab", "link"].indexOf(value) !== -1;
      }
    },
    ariaHasPopup: {
      type: String,
      default: null,
      validator: function(value) {
        // Based on https://labs.levelaccess.com/index.php/ARIA_Haspopup_property
        return (
          ["menu", "listbox", "tree", "grid", "dialog", "true"].indexOf(
            value
          ) !== -1
        );
      }
    },
    ariaRoledescription: {
      type: String,
      default: null,
      validator: function(value) {
        // Based on https://www.w3.org/TR/wai-aria-1.1/#aria-roledescription
        return value.replace(/\s/g, "").length > 0;
      }
    },
    disabled: {
      type: Boolean,
      default: false
    }
  },
  computed: {
    classes() {
      return "PButton " + this.cClass;
    }
  },
  emits: ["click", "keydown"],
  methods: {
    handleClick() {
      this.$emit("click");
    },
    handleKeyDown() {
      this.$emit("keydown");
    }
  }
};
</script>

<style lang="less" scoped>
@import "../../stylesheets/shadows.less";

.PButton {
  background-color: inherit;
  border: 1px solid #dedede;
  cursor: pointer;
  padding: 1rem;
  border-radius: 1rem;
  font-size: 1rem;
}

.PButton:disabled,
.PButton[disabled] {
  background-color: #e0e0e0;
  box-shadow: none;
  cursor: not-allowed;
  color: #484747;
}

// Cool provided styling
.PButton--solid-box-shadow {
  transition: box-shadow 0.2s, background-color 0.2s;

  &:hover,
  &:focus {
    box-shadow: 0 0 0 0.4rem rgba(0, 0, 0, 0.1);
  }

  &:active {
    background: #f8f8f8;
    box-shadow: 0 0 0 0.2rem rgba(0, 0, 0, 0.1);
  }
}

.PButton--elevated {
  transition: box-shadow 0.2s;
  box-shadow: @box-shadow-01;

  &:hover,
  &:focus {
    box-shadow: @box-shadow-02;
  }

  &:active {
    box-shadow: @box-shadow-01;
  }
}

.PButton--stealth {
  border: none;
  transition: background-color 0.2s;

  &:hover,
  &:focus {
    background-color: #f2f2f2;
  }

  &:active {
    background-color: #d9d9d9;
  }
}
</style>
