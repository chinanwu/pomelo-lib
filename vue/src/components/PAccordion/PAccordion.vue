<template>
  <div>
    <PButton
      id="pAccordionBtn"
      class="PAccordionBtn"
      @click="expanded = !expanded"
      @keydown="expanded = !expanded"
    >
      <template #content>
        <span>{{ label }}</span>
        <span>{{ expanded ? "-" : "+" }}</span>
      </template>
    </PButton>
    <transition name="expand" mode="out-in">
      <slot v-if="expanded" name="content"></slot>
    </transition>
  </div>
</template>

<script>
import PButton from "@/components/PButtons/PButton";

export default {
  name: "PAccordion",
  components: {
    PButton
  },
  props: {
    label: {
      type: String,
      default: null
    },
    expand: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      expanded: this.expand
    };
  },
  methods: {
    handleKeyDown(event) {
      if (
        event &&
        event.keyCode &&
        !event.shiftKey &&
        !event.ctrlKey &&
        !event.altKey &&
        !event.metaKey &&
        (event.keyCode === 13 || event.keyCode === 32) // if enter or space is pressed
      ) {
        event.preventDefault();
        this.expanded = !this.expanded;
      }
    }
  }
};
</script>

<style lang="less" scoped>
.PAccordionBtn {
  width: 100%;
  text-align: left;
  padding: 1rem;
  background-color: #efefef;
  border: none;
  display: flex;
  justify-content: space-between;
  cursor: pointer;
  transition: background-color 0.2s;
  border-radius: 0.5rem;
  font-size: 1.4rem;

  &:hover,
  &:focus {
    background-color: #f4f4f4;
  }

  &:active {
    background-color: #efefef;
  }
}

.expand {
  &-enter-active,
  &-leave-active {
    transition: all 0.3s;
    overflow: hidden;
  }
  &-enter-from,
  &-leave-to {
    max-height: 0;
    opacity: 0;
  }
  &-enter-to,
  &-leave-from {
    max-height: 100vh;
    opacity: 1;
  }
}
</style>
