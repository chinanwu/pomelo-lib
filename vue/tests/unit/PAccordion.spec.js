import { mount, shallowMount } from "@vue/test-utils";

import PAccordion from "@/components/PAccordion/PAccordion.vue";
import PButton from "@/components/PButtons/PButton.vue";

describe("PAccordion", () => {
  describe("rendering", () => {
    it("renders PAccordion", () => {
      const wrapper = shallowMount(PAccordion);
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.props().expand).toBe(false);
    });

    it("renders expanded accordion with label", () => {
      const wrapper = mount(PAccordion, {
        props: {
          label: "Accordion Label",
          expand: true
        },
        slots: {
          content: "I am the innards of the accordion"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe("behaviour", () => {
    it("expands on click", () => {
      const wrapper = shallowMount(PAccordion, {
        props: {
          expand: false
        }
      });
      wrapper.findComponent(PButton).vm.$emit("click");
      expect(wrapper.vm.expanded).toEqual(true);
    });

    it("closes on click", () => {
      const wrapper = shallowMount(PAccordion, {
        props: {
          expand: true
        }
      });
      wrapper.findComponent(PButton).vm.$emit("click");
      expect(wrapper.vm.expanded).toEqual(false);
    });
  });
});
