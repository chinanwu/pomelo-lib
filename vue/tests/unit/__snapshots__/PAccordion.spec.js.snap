// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PAccordion rendering renders PAccordion 1`] = `
<div>
  <p-button-stub
    class="PAccordionBtn"
  />
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`PAccordion rendering renders expanded accordion with label 1`] = `
<div>
  <button
    class="PButton PAccordionBtn"
    id="pAccordionBtn"
  >
    
    <span>
      Accordion Label
    </span>
    <span>
      -
    </span>
    
  </button>
  <transition-stub>
    
    
    I am the innards of the accordion
    
    
  </transition-stub>
</div>
`;
